### DISCUSSION TITLE
New features and other changes for 23.04.0

### DISCUSSION THREAD

**Discussion by tadeubas** [selfcustody/krux#202] [Show and tell]
@jreesun , @odudex I made the following changes to my `new_features` branch, what do you think would be nice to have in the next Krux version?

1. Fixed and created some missing **translations**
2. Added more info to **README**
3. Customized `input.wait_for_press()` to allow messages to be presented for some time or to be discarded by user input
4. **CategorySetting** now support colors! (Used already in **Logging** and **Bitcoin**)
5. **Status bar logging_indicator** now shows color according to selected log level
6. **Status bar** now shows a `test` indicator in green when on **testnet**
7. Created a None Printer Driver in Settings as default, in order to avoid `Print?` prompt to users that don't have a printer
8. Removed the `Esc` key prompt `Are you sure?` in Settings
9. Fixed some `Missing attribute errors` that occurred when printing with `cnc/file` Printer Driver. We also now avoid printing text to this Driver (currently it doesn't support)
10. `Print?` and `Print to QR?` prompt now shows the current selected Printer Driver (the prompt don't appear if the selected Driver is none)
11. ~~Added **Fingerprint and Derivation path** information to `Home> Mnemonic> Words` screen~~
12. Allowed to skip with `ENTER/TOUCH` on QRCodes presented at `Home> Extended Public Key` (please see: https://github.com/selfcustody/krux/discussions/200 )
13. Allowed to skip with `ENTER/TOUCH` on QRCodes presented at `Home> Mnemonic> Plaintext QR` (please see: https://github.com/selfcustody/krux/discussions/200 )
14. Added a `Checking for SD card` message right before `SDCard.remount()` (like we do with `Loading Camera`)
15. Created two new ways to import Words from digits (Hexadecimal and Octal)
16. Created new test cases and refactored some to increase readability and simplicity
17. Other minor changes or fixes
18. Added Back option to some submenus
19. New feature: Address List #194 
20. Passphrase ESC key don't discard the entered mnemonic anymore, user is enforced to either click NO on passphrase prompt or define a passphrase (if clicked on YES for passphrase)
21. Now a fingerprint confirm prompt appears right before the home screen
22. New feature: SD file explorer (to select the file to sign) and SD filename creator (to select the name of the file to save) for Sign PSBT and Sign Message
23. Now the derivation path is shown in the selection screen along with Single-key or Multisig options
24. Sign PSBT shows more details before signing
25. Created new language translation: pokemon
26. New setting: Persist (choose where to persist settings. flash or sd [flash is now the default])
27. ~~New menu item: SD Check (will inform if SD is detected and the size / used / free)~~
28. Font: added ₽ symbol for Pokémon Dollar (customized with 2 lines like Yen), refactored ₿ glyph
29. New menu item: Tools with sub-menu `Check SD Card`, `Print Test Page` and `Create QR Code`
30. Now settings show a warn message when user set a `NumberSettings` that is out of range
31. Removed dot (`.`) from `NumberSettings` when type is `int`
32. Passphrase now accepts scanning of a QR code
33. ~~Sign PSBT always show QR Code in FORMAT_PMOFN~~

**Comment by odudex** [comment]
I built your branch here, for me all changes are good! Nice work on tests, translations, readme!
Would just recommend test carefully all types of QRs, including animated on changes 12 and 13(I didn't tested it yet).
I find load as hex digits cool, saw some metal plates that use it. As octal I think has fewer use, but I'm ok with it.

**Comment by ghost** [comment]
1 through 10 sound like small and good changes. 9 was already on my personal todo list since I knew that was going to cause problems with the CNC driver, so I'm glad you were able to recognize and fix that!

7 is an elegant solution, nice!

11 sounds strange to me since the mnemonic has nothing to do with the fingerprint or derivation, and that information is already displayed under the Extended Public Key pages. What's the reasoning here for adding them to the words page?

For 12 and 13 I'll need to do some checking to make sure this doesn't break anything. The reason pages with QR codes behave differently than other pages from a UI/UX perspective is because you used to (or maybe still can, I don't remember) be able to manually page through an animated QR code with the PAGE/PAGE_PREV buttons.

14, 16, 17 sound fine. 👍 

15 should land in the next release, if at all, since it's just a new feature entirely and not fixing anything. But I have to ask: do people actually backup their mnemonics in hex or octal? I haven't heard of it. If they do, then I don't mind adding it, otherwise it feels like adding a feature just because we can. My preference would be to add support for Word Indexes (ie, 0-based indexing) before Hex and Octal since [I know some people want to do it that way](https://github.com/SeedSigner/seedsigner/issues/211#issuecomment-1137208528).

**Reply by ghost** [reply]
Missed @odudex's  reply when I was drafting mine:
>  I find load as hex digits cool, saw some metal plates that use it. As octal I think has fewer use, but I'm ok with it.

If that's the case, then I'm okay with adding hex (in the next release), but let's punt on octal until someone needs it.

**Reply by tadeubas** [reply]
I forgot to mention the comment here: https://github.com/SeedSigner/seedsigner/issues/211#issuecomment-1137208528

I've changed the `possible_letters` for "Words > From Digits", now the `0` appears disabled for the first number of a word. Krux also says that it is importing a word index from 1 to 2048 and the user needs to press ok before importing. Krux always show the word the user selected and at the end all words asking for confirmation... so I don't see any problem here as there exists metal wallets that asks to the user to bkp the word index (or encourage doing this).

EDIT: Only developers known that list indexes start from 0 on computers. BIP39, list of address from wallets and every other software (Excel or LibreOffice) starts with index 1, this is the most common for users

**Comment by odudex** [comment]
One hex storage [example ](https://web.archive.org/web/20210118061112/https://www.secretxseed.com/products/enigma-card)

**Comment by tadeubas** [comment]
> 11 sounds strange to me since the mnemonic has nothing to do with the fingerprint or derivation, and that information is already displayed under the Extended Public Key pages. What's the reasoning here for adding them to the words page?

These are important information, when the user imports the words and confirms clicking on Single-Key for example, he doesn't know what type of wallet has been imported... it can be:

- BIP32 m/0'/0'
- m/84'/0'/0'/0
- m/49'/0'/0'/0
- m/44'/0'/0'/0

So I wanted to present this information somewhere... I didn't want to create a new menu just for this, so I assume the user would usually check the mnemonic to confirm what Krux had imported, and for Dock and Amigo, this screen has enough space to show this info (however, m5stick needs one more screen). I know the information is presented at `Extended Public Key` (XPUB) screen, but it is confusing to me to see this info there, because XPUB generally means only a public key, an information that starts with XPUB/YPUB/ZPUB and not the Master fingerprint or the Derivation path. BW for example on the Wallet menu screen has a button "Show Wallet XPUB" and only gives the zpub info... some other wallets do the same! It was strange to me to see this info there, and I know others will make the same confusion, so i thought, why not present this info on mnemonic screen too? If we remove from there, I think we should add some kind of "Wallet info" menu item

**Reply by odudex** [reply]
@tadeubas the fingerprint and derivation are always exported together with the xpub/zpub as a "header"  in the QR code data, and used by coordinators to define the script type and "ID" of the the wallet .
Ex:  This is a xpub QR code data:
[3f635a63/84h/0h/0h]xpub6CYYYw6h668PkCSXxH9yxBG32zCMEb6N9DuVY8Ax8U7RSV86qKrrhjJfS6nL5jSoikLpd1Qw9qgHv5vyRi7V4nfV3ymLfGpFShsYsFmQiT8

**Reply by tadeubas** [reply]
Sure, but this is the way Krux works... BW and others only shows the xpub/zpub when talking about `Extended public key`... so it would need some time on Krux to realize where this info is. I was just explaining my motivations and idea behind exposing this info on two different screens, if you want this to be removed, no problem!

**Reply by tadeubas** [reply]
SeedSigner shows the fingerprint right after confirming words, just before setting a passphrase and I think it shows the fingerprint after setting the passphrase too... @jreesun @odudex what do you guys think about implementing this also in Krux?
![image](https://user-images.githubusercontent.com/5518496/222707968-92891acc-6e21-4009-9669-d00caa05258f.png)

**Reply by odudex** [reply]
> Sure, but this is the way Krux works... 

I think Specter, at least, also works like this.
Specter has been used as a reference more frequently than Seedsigner on Krux project.

**Reply by tadeubas** [reply]
nice to know... I don't have a Specter either 😟 But what about the fingerprint screen confirmation after the seed confirmation? SeedSigner does the same after BIP-39 Passphrase is set
![verify-passphrase](https://user-images.githubusercontent.com/5518496/222730409-8e6d59ad-30b8-47ae-8a5e-92ca1f1dcadb.PNG)

And later it shows the fingerprint at the top all the time (I think it is because they can load up to 3 seeds)
![menu-seed-mostra-fingerprint](https://user-images.githubusercontent.com/5518496/222730681-e9666ce2-0fd1-42e4-a26d-84ea4e612340.PNG)
![image](https://user-images.githubusercontent.com/5518496/222730892-8f9fc366-2a8f-4cf1-a59a-2dc6e9a69a17.png)

**Comment by tadeubas** [comment]
For 12 and 13 I didn't see any problem, but will test it better on the m5stick simulator (I just have a maix dock)

**Reply by tadeubas** [reply]
@jreesun and @odudex I tested **12** and **13**, mnemonic 12w/24w, single/multi, with imported wallet descriptor or without... not once `Home> Extended Public Key` or `Home> Mnemonic> Plaintext QR` displayed a QR code other than 1/1 on the simulator (maybe it's different on the device?). Can anyone test this too? I didn't see any problems, but I didn't see any animated QRs on either of those screens
![image](https://user-images.githubusercontent.com/5518496/222595065-d1762c67-b5ba-4723-855b-2ac7ac7281fb.png)

**Reply by odudex** [reply]
The animated QRs are used to show signed transactions, so you have to test the transaction signing workflow. Hopefully I'll be able test tomorrow too, will let you know how it went.

**Reply by tadeubas** [reply]
I haven't changed the way QRCodes work on other screens, to accept ENTER/TOUCH yet... just `Home> Extended Public Key` and `Home> Mnemonic> Plaintext QR`. It is what I say on **12. and 13.**

**Reply by odudex** [reply]
> I haven't changed the way QRCodes work on other screens, to accept ENTER/TOUCH yet... just `Home> Extended Public Key` and `Home> Mnemonic> Plaintext QR`. It is what I say on **12. and 13.**

@tadeubas I think you did change code shared with animated QR code, ex: `display_qr_codes` function. So it is recommended to test all features that use it. I say it because I have broken animated QR codes when making static QR code and input modifications before. Your code however is working, I've tested it now.

**Comment by tadeubas** [comment]
Added items 18 and 19

**Comment by tadeubas** [comment]
Removed `11. Added Fingerprint and Derivation path information to Home> Mnemonic> Words screen` in favor of the `21. Now a fingerprint confirm prompt appears right before the home screen`

**Comment by tadeubas** [comment]
Added items 20. and 21.

**Comment by tadeubas** [comment]
Added item 22. `New feature: SD file explorer (to select the file to sign) and SD filename creator (to select the name of the file to save) for Sign PSBT and Sign Message`

**Comment by tadeubas** [comment]
Added item 23. `Now the derivation path is shown in the selection screen along with Single-key or Multisig options`

**Comment by tadeubas** [comment]
Added item 24. `Sign PSBT shows more details before signing`

Example of a small transaction:
![image](https://user-images.githubusercontent.com/5518496/224578608-a850cf38-22ee-47d6-957c-d8d43c7e4557.png)

Example of a big transaction (part1 and part 2):
![image](https://user-images.githubusercontent.com/5518496/224578628-b7c8ebb2-4bf8-4d05-83d0-6ade507f1dad.png)
![image](https://user-images.githubusercontent.com/5518496/224578645-93832269-9601-4217-a336-25dc1126f79d.png)

**Comment by tadeubas** [comment]
Added item 25. `Created new language translation: pokemon`. It is basic english but with the modifications below (useful for op-sec and 4fun 😄):
* BIP39 to Pokédex
* Words to Pokémons
* Mnemonic to Names
* Wallet to Team
* Tiny Seed to Tiny Card
* SeedQR to CardQR
* Stackbit 1248 to Poké Stack 1248
* Network to Version
* Krux to Pokémon
* Key to Pokéball
* Signature to Enroll
* Address to League
* PSBT to Creature

Also the SPLASH changes to an ASCII Pikachu!

**Comment by tadeubas** [comment]
Added items 26. `New setting: Persist (choose where to persist settings. flash or sd [flash is now the default])` and 27. `New menu item: SD Check (will inform if SD is detected and the size / used / free)`

![image](https://user-images.githubusercontent.com/5518496/225708772-bce1e5da-3568-4d3e-b6f1-cfc82f48c099.png)

**Reply by odudex** [reply]
About flash usage, that would make Krux not be stateless, so some people wouldn't like it. Maybe it shouldn't be default and have some kind of warning when activated?

I would create a main menu item called "Tools", in which these features (to be built) could also be added:
- Check SD Card
- Print Test Page
- Create QR from text (to optionally be used as QR passphrase)

**Reply by tadeubas** [reply]
I agree!

**Reply by tadeubas** [reply]
Removed item 26. in favor of item 29. `New menu item: Tools with sub-menu Check SD Card, Print Test Page and Create QR Code`

**Comment by tadeubas** [comment]
Added item 28. `Font: added ₽ symbol for Pokémon Dollar (customized with 2 lines like Yen), refactored ₿ glyph`

**Reply by tadeubas** [reply]
This item needs this PR from MaixPy project: https://github.com/selfcustody/MaixPy/pull/6

**Comment by tadeubas** [comment]
Added items 30. and 31.

**Comment by tadeubas** [comment]
Added item 32. `Passphrase now accepts scanning of a QR code`

**Comment by tadeubas** [comment]
Added item 33. `Sign PSBT always show QR Code in FORMAT_PMOFN`. This is a big improvement because QR codes in `FORMAT_UR` have 4.8 times the data size compared to `FORMAT_PMOFN`. Now signing the same PSBT via camera (QR reader) or via SD file gives the same QR code output on screen and printer

**Reply by tadeubas** [reply]
Removed item 33. because BW doesn't accept `FORMAT_PMOFN`
